### PR TITLE
Fixes in image import controller

### PR DESCRIPTION
**Pull request by soltysh**
I've reworked the the error reporting from `getTags` and `imprortTags` to include information whether an error is or not a retry-able. The former should just return the error and retry controller will re-kick the import. The latter should place the error in an import annotation. 

Additionally I've reworked slightly how we print tags in our describer, to show all the tags from spec and status and inform about status of each of those. This is needed for external tags to be printed.

This fixes problem described in https://bugzilla.redhat.com/show_bug.cgi?id=1275537

@smarterclayton @miminar ptal

### CODE REVIEW

**Comment by soltysh** [comment]
[test]

**Comment by soltysh** [comment]
[test]

**Comment by soltysh** [comment]
I've added one more v(5).log in the `importTag` method.

File: pkg/image/controller/controller.go [hunk 1]
```diff
@@ -67,19 +66,20 @@ func (c *ImportController) Next(stream *api.ImageStream) error {
 		client = dockerregistry.NewClient()
 	}
 
-	toImport, err := getTags(stream, client, insecure)
+	toImport, err, retry := getTags(stream, client, insecure)
```
Comment (miminar) [inline pkg/image/controller/controller.go]: Shouldn't we at least log `err` if `len(toImport) > 0`?

File: pkg/image/controller/controller.go [hunk 2]
```diff
@@ -87,9 +87,10 @@ func (c *ImportController) Next(stream *api.ImageStream) error {
 }
 
 // getTags returns tags from default upstream image repository and explicitly defined.
-// Returns a map of tags to be imported and an error if one occurs.
+// Returns a map of tags to be imported, an error if one occurs and flag saying if
```
Comment (miminar) [inline pkg/image/controller/controller.go]: Above two lines could be merged.

File: pkg/image/controller/controller.go [hunk 3]
```diff
@@ -87,9 +87,10 @@ func (c *ImportController) Next(stream *api.ImageStream) error {
...
+// we should retry imports, meaning not setting the import annotation.
 // Tags explicitly defined will overwrite those from default upstream image repository.
-func getTags(stream *api.ImageStream, client dockerregistry.Client, insecure bool) (map[string]api.DockerImageReference, error) {
+func getTags(stream *api.ImageStream, client dockerregistry.Client, insecure bool) (map[string]api.DockerImageReference, error, bool) {
```
Comment (smarterclayton) [inline pkg/image/controller/controller.go]: Change the signatures of all of these methods so that error is last (error is always last)

**Comment by soltysh** [comment]
@smarterclayton @miminar I've addressed your comments:
1. changed method signatures to have error as the last element
2. added aggregation logic for errors, so that user is always informed of all the errors in the annotation (no hidden errors)
3. changed the if-retry order

The only thing I didn't change was the `getTags` error checking @smarterclayton asked, my reasoning are [here](https://github.com/soltysh/origin/commit/e6a3f0b7f2a142fcdd6e0ad3990d993b3e8ac921#commitcomment-14147879). ptal

**Comment by soltysh** [comment]
Oh yes, and I've merged the double comments, @miminar pointed out.

**Comment by miminar** [comment]
LGTM

File: pkg/image/controller/controller.go [hunk 4]
```diff
@@ -67,29 +66,38 @@ func (c *ImportController) Next(stream *api.ImageStream) error {
 		client = dockerregistry.NewClient()
 	}
 
-	toImport, err := getTags(stream, client, insecure)
+	var errlist []error
+	toImport, retry, getErr := getTags(stream, client, insecure)
```
Comment (smarterclayton) [inline pkg/image/controller/controller.go]: Why are you using `getErr` instead of `err`?

File: pkg/image/controller/controller.go [hunk 5]
```diff
@@ -67,29 +66,38 @@ func (c *ImportController) Next(stream *api.ImageStream) error {
...
+	if getErr != nil {
+		errlist = append(errlist, getErr)
 	}
 
-	errs := c.importTags(stream, toImport, client, insecure)
-	// one of retry-able error happened, we need to inform the RetryController
-	// the import should be retried by returning error
-	if len(errs) > 0 {
-		return kerrors.NewAggregate(errs)
+	retry, importErr := c.importTags(stream, toImport, client, insecure)
```
Comment (smarterclayton) [inline pkg/image/controller/controller.go]: Should be `err` instead of `importErr`

File: pkg/image/controller/controller_test.go [hunk 6]
```diff
@@ -606,6 +606,15 @@ func TestControllerWithSpecTags(t *testing.T) {
 						},
 					},
 				},
+				{
```
Comment (smarterclayton) [inline pkg/image/controller/controller_test.go]: You need test cases verifying that retries are being returned by those methods when you expect.  I'd like to see much more on that before we merge this.

**Comment by smarterclayton** [comment]
This is P0 but I can't merge this without more tests validating retries are happening.  Can you get those tests added in the next two hours?

File: pkg/image/controller/controller.go (hunk 4 above)
Comment (soltysh) [inline pkg/image/controller/controller.go]: Right, forgot to refactor that, I had some others ideas about merging the two errors and needed different names then.

**Comment by soltysh** [comment]
@smarterclayton on it.

**Comment by soltysh** [comment]
@smarterclayton updated and added testcases as requested, ptal

**Comment by smarterclayton** [comment]
[merge]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[viptest]

**Comment by smarterclayton** [comment]
[viptest]

**Comment by smarterclayton** [comment]
[viptest]

**Comment by smarterclayton** [comment]
This passed all the way to the end of e2e and then failed.

**Comment by smarterclayton** [comment]
I'm seeing a tag that doesn't exist (via a direct tag reference) never result in failure.  

```
apiVersion: v1
items:
- apiVersion: v1
  kind: ImageStream
  metadata:
    creationTimestamp: 2015-11-04T19:36:28Z
    name: deploy-tester
    namespace: default
    resourceVersion: "535"
    selfLink: /oapi/v1/namespaces/default/imagestreams/deploy-tester
    uid: 58046fd9-832b-11e5-bb0d-080027c5bfa9
  spec:
    dockerImageRepository: ironcladlou/deploy-tester
    tags:
    - from:
        kind: DockerImage
        name: kubernetes/guestbook:latest
      name: latest
  status:
    dockerImageRepository: ironcladlou/deploy-tester
kind: List
metadata: {}
```

retries forever because `kubernetes/guestbook:latest` doesn't exist.  Can you pull together a fix for 3.1.1?

**Pull request merged**